
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Week 11: Genetic drift and simulating the Buri experiment &#8212; BIOL301 Fall 2022 materials</title>
    
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">
<link href="_static/styles/pydata-sphinx-theme.css?digest=1999514e3f237ded88cf" rel="stylesheet">

    
  <link rel="stylesheet"
    href="_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" href="_static/styles/sphinx-book-theme.css?digest=5115cc725059bd94278eecd172e13a965bf8f5a9" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="_static/design-style.b7bb847fb20b106c3d81b95245e65545.min.css" />
    
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf">

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script src="_static/scripts/sphinx-book-theme.js?digest=9c920249402e914e316237a7dbc6769907cce411"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Week 10: Master equations applied to gene expression" href="10_master_equation_gene_exp.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="60">
<!-- Checkboxes to toggle the left sidebar -->
<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation" aria-label="Toggle navigation sidebar">
<label class="overlay overlay-navbar" for="__navigation">
    <div class="visually-hidden">Toggle navigation sidebar</div>
</label>
<!-- Checkboxes to toggle the in-page toc -->
<input type="checkbox" class="sidebar-toggle" name="__page-toc" id="__page-toc" aria-label="Toggle in-page Table of Contents">
<label class="overlay overlay-pagetoc" for="__page-toc">
    <div class="visually-hidden">Toggle in-page Table of Contents</div>
</label>
<!-- Headers at the top -->
<div class="announcement header-item noprint"></div>
<div class="header header-item noprint"></div>

    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<!-- Sidebar -->
<div class="bd-sidebar noprint" id="site-navigation">
    <div class="bd-sidebar__content">
        <div class="bd-sidebar__top"><div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="_static/download.jpg" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">BIOL301 Fall 2022 materials</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="intro.html">
                    Welcome to the BIOL301 Webpage
                </a>
            </li>
        </ul>
        <ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="01_configuring_computer.html">
   Week 1: Configuring your computer
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="01_intro_to_jupyter.html">
   Week 1: Introduction to JupyterLab
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="01_terminal_commands.html">
   Week 1: Terminal Commands
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="01_intro_to_latex.html">
   Week 1: Introduction to LaTeX
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="03_numerical_integration.html">
   Week 3: Numerical Integration
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="03_slope_fields.html">
   Week 3: Slope fields and other graphical analysis
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="04_lotka_volterra.html">
   Week 4: Lotka-Volterra
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="04_scipy_integration.html">
   Week 4: Numerical integration with scipy
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="05_bifurcations.html">
   Week 5: Bifurcations
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="05_midterm1_review.html">
   Midterm 1 Review
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="06_insect.html">
   Week 6: Insect outbreak with interactive plots
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="07_gene_expression.html">
   Week 7: Modeling mRNA and protein production
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="08_stochastic_simulation.html">
   Week 8: Stochastic simulations – Gillespie Algorithm
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="09_master_equation.html">
   Week 9: Numerically integrating a probability distribution – Master equations
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="10_master_equation_gene_exp.html">
   Week 10: Master equations applied to gene expression
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Week 11: Genetic drift and simulating the Buri experiment
  </a>
 </li>
</ul>

    </div>
</nav></div>
        <div class="bd-sidebar__bottom">
             <!-- To handle the deprecated key -->
            
            <div class="navbar_extra_footer">
            Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
            </div>
            
        </div>
    </div>
    <div id="rtd-footer-container"></div>
</div>


          


          
<!-- A tiny helper pixel to detect if we've scrolled -->
<div class="sbt-scroll-pixel-helper"></div>
<!-- Main content -->
<div class="col py-0 content-container">
    
    <div class="header-article row sticky-top noprint">
        



<div class="col py-1 d-flex header-article-main">
    <div class="header-article__left">
        
        <label for="__navigation"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="right"
title="Toggle navigation"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-bars"></i>
  </span>

</label>

        
    </div>
    <div class="header-article__right">
<div class="menu-dropdown menu-dropdown-launch-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Launch interactive content">
      <i class="fas fa-rocket"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://mybinder.org/v2/gh/sbeeler/BIOL301_Fall2022/master?urlpath=tree/docs/11_genetic_drift.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Launch on Binder"
>
  

<span class="headerbtn__icon-container">
  
    <img src="_static/images/logo_binder.svg">
  </span>
<span class="headerbtn__text-container">Binder</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<button onclick="toggleFullScreen()"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="bottom"
title="Fullscreen mode"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>

<div class="menu-dropdown menu-dropdown-repository-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Source repositories">
      <i class="fab fa-github"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="https://github.com/sbeeler/BIOL301_Fall2022/tree/gh-pages"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Source repository"
>
  

<span class="headerbtn__icon-container">
  <i class="fab fa-github"></i>
  </span>
<span class="headerbtn__text-container">repository</span>
</a>

      </li>
      
      <li>
        <a href="https://github.com/sbeeler/BIOL301_Fall2022/tree/gh-pages/issues/new?title=Issue%20on%20page%20%2F11_genetic_drift.html&body=Your%20issue%20content%20here."
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Open an issue"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="headerbtn__text-container">open issue</span>
</a>

      </li>
      
    </ul>
  </div>
</div>

<div class="menu-dropdown menu-dropdown-download-buttons">
  <button class="headerbtn menu-dropdown__trigger"
      aria-label="Download this page">
      <i class="fas fa-download"></i>
  </button>
  <div class="menu-dropdown__content">
    <ul>
      <li>
        <a href="_sources/11_genetic_drift.ipynb"
   class="headerbtn"
   data-toggle="tooltip"
data-placement="left"
title="Download source file"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="headerbtn__text-container">.ipynb</span>
</a>

      </li>
      
      <li>
        
<button onclick="printPdf(this)"
  class="headerbtn"
  data-toggle="tooltip"
data-placement="left"
title="Print to PDF"
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="headerbtn__text-container">.pdf</span>
</button>

      </li>
      
    </ul>
  </div>
</div>
<label for="__page-toc"
  class="headerbtn headerbtn-page-toc"
  
>
  

<span class="headerbtn__icon-container">
  <i class="fas fa-list"></i>
  </span>

</label>

    </div>
</div>

<!-- Table of contents -->
<div class="col-md-3 bd-toc show noprint">
    <div class="tocsection onthispage pt-5 pb-3">
        <i class="fas fa-list"></i> Contents
    </div>
    <nav id="bd-toc-nav" aria-label="Page">
        <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#simulating-a-single-vial-of-a-buri-experiment">
   Simulating a single vial of a Buri experiment
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#problem-1-simulating-the-full-buri-experiment">
   Problem 1 – Simulating the full Buri experiment
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#problem-2-modifying-the-population-size">
   Problem 2 – Modifying the population size
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#problem-3-modifying-the-number-of-generations">
   Problem 3 – Modifying the number of generations
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#problem-4-going-beyond-bag-of-alleles">
   Problem 4 – Going beyond “bag of alleles”
  </a>
 </li>
</ul>

    </nav>
</div>
    </div>
    <div class="article row">
        <div class="col pl-md-3 pl-lg-5 content-container">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>Week 11: Genetic drift and simulating the Buri experiment</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#simulating-a-single-vial-of-a-buri-experiment">
   Simulating a single vial of a Buri experiment
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#problem-1-simulating-the-full-buri-experiment">
   Problem 1 – Simulating the full Buri experiment
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#problem-2-modifying-the-population-size">
   Problem 2 – Modifying the population size
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#problem-3-modifying-the-number-of-generations">
   Problem 3 – Modifying the number of generations
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#problem-4-going-beyond-bag-of-alleles">
   Problem 4 – Going beyond “bag of alleles”
  </a>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            <main id="main-content" role="main">
                
              <div>
                
  <section class="tex2jax_ignore mathjax_ignore" id="week-11-genetic-drift-and-simulating-the-buri-experiment">
<h1>Week 11: Genetic drift and simulating the Buri experiment<a class="headerlink" href="#week-11-genetic-drift-and-simulating-the-buri-experiment" title="Permalink to this headline">#</a></h1>
<hr><div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">numba</span>
<span class="kn">import</span> <span class="nn">random</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">scipy.integrate</span>

<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
<span class="n">sns</span><span class="o">.</span><span class="n">set</span><span class="p">(</span><span class="n">font_scale</span> <span class="o">=</span> <span class="mf">1.5</span><span class="p">,</span> <span class="n">rc</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;lines.linewidth&quot;</span><span class="p">:</span><span class="mi">2</span><span class="p">})</span>
</pre></div>
</div>
</div>
</div>
<section id="simulating-a-single-vial-of-a-buri-experiment">
<h2>Simulating a single vial of a Buri experiment<a class="headerlink" href="#simulating-a-single-vial-of-a-buri-experiment" title="Permalink to this headline">#</a></h2>
<p>Let’s simulate how a single vial might drift over time in the Buri experiment. We will first initialize some parameters:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># number of flies in each vial</span>
<span class="n">n_flies</span> <span class="o">=</span> <span class="mi">16</span>

<span class="c1"># number of generations to run the simulation</span>
<span class="n">n_generations</span> <span class="o">=</span> <span class="mi">20</span>

<span class="c1"># initilaize an array to store the frequency of the red allele</span>
<span class="n">r_freq</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">(</span><span class="n">n_generations</span><span class="p">)</span>
<span class="n">r_freq</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="mf">0.5</span>
<span class="n">r_freq</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>array([0.5, 0. , 0. , 0. , 0. , 0. , 0. , 0. , 0. , 0. , 0. , 0. , 0. ,
       0. , 0. , 0. , 0. , 0. , 0. , 0. ])
</pre></div>
</div>
</div>
</div>
<p>How will the allele frequency change over time for our vial? Using the “bag of alleles” approximation, we will draw the next generation’s 32 alleles, each with a <code class="docutils literal notranslate"><span class="pre">0.5</span></code> chance of being red. For this we can use <code class="docutils literal notranslate"><span class="pre">np.random.binomial(n,</span> <span class="pre">p)</span></code>, where <code class="docutils literal notranslate"><span class="pre">n</span></code> will be the number of alleles we want and <code class="docutils literal notranslate"><span class="pre">p</span></code> will be the probability of grabbing a red allele out of our “bag”.</p>
<p>As an example:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">n</span> <span class="o">=</span> <span class="n">n_flies</span> <span class="o">*</span> <span class="mi">2</span>
<span class="n">p</span> <span class="o">=</span> <span class="mf">0.5</span>

<span class="c1"># how many red alleles do we get?</span>
<span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">binomial</span><span class="p">(</span><span class="n">n</span><span class="p">,</span> <span class="n">p</span><span class="p">)</span> 
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>15
</pre></div>
</div>
</div>
</div>
<p>If you run that cell repeatedly, you’ll see that the average result will be that 16 alleles will be red, but this value can vary.</p>
<p>Using this approach for generating the next generation’s red allele frequency, let’s simulate a single vial over time, updating the frequency at each generation:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">n_generations</span><span class="p">):</span>
    <span class="n">p</span> <span class="o">=</span> <span class="n">r_freq</span><span class="p">[</span><span class="n">i</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
    <span class="n">num_red</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">binomial</span><span class="p">(</span><span class="n">n_flies</span><span class="o">*</span><span class="mi">2</span><span class="p">,</span> <span class="n">p</span><span class="p">)</span> 
    <span class="n">new_r_freq</span> <span class="o">=</span> <span class="n">num_red</span> <span class="o">/</span> <span class="p">(</span><span class="n">n_flies</span><span class="o">*</span><span class="mi">2</span><span class="p">)</span>
    <span class="n">r_freq</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">new_r_freq</span>

<span class="n">r_freq</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>array([0.5    , 0.5    , 0.65625, 0.65625, 0.65625, 0.75   , 0.78125,
       0.78125, 0.875  , 0.875  , 0.875  , 0.84375, 0.9375 , 0.90625,
       0.96875, 0.9375 , 0.96875, 1.     , 1.     , 1.     ])
</pre></div>
</div>
</div>
</div>
<p>Plotting the results. Does the red or white allele become fixed?</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">plt</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">r_freq</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">xlabel</span><span class="p">(</span><span class="s2">&quot;generation&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylabel</span><span class="p">(</span><span class="s2">&quot;red allele frequency&quot;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">ylim</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>(0.0, 1.0)
</pre></div>
</div>
<img alt="_images/11_genetic_drift_9_1.png" src="_images/11_genetic_drift_9_1.png" />
</div>
</div>
</section>
<section id="problem-1-simulating-the-full-buri-experiment">
<h2>Problem 1 – Simulating the full Buri experiment<a class="headerlink" href="#problem-1-simulating-the-full-buri-experiment" title="Permalink to this headline">#</a></h2>
<p>Extend what we’ve done so far to model the entire Buri experiment, tracking 107 vials over 20 generations. For this, we will want to use a 2D-array, such as <code class="docutils literal notranslate"><span class="pre">red_freq</span> <span class="pre">=</span> <span class="pre">np.zeros([n_vials,</span> <span class="pre">n_generations])</span></code> to store our allele frequencies in.</p>
<p>Plot the distribution of the red allele frequency at the end of the 20 generations as a histogram. It should look something like this:</p>
<p><img alt="Screen Shot 2022-11-02 at 11.37.20 PM.png" src="attachment:6a35705e-e25f-4507-a124-2e7a9ae1620f.png" /></p>
</section>
<section id="problem-2-modifying-the-population-size">
<h2>Problem 2 – Modifying the population size<a class="headerlink" href="#problem-2-modifying-the-population-size" title="Permalink to this headline">#</a></h2>
<p>Rerun your simulation for <code class="docutils literal notranslate"><span class="pre">n_flies</span> <span class="pre">=</span> <span class="pre">8</span></code> and <code class="docutils literal notranslate"><span class="pre">n_flies</span> <span class="pre">=</span> <span class="pre">32</span></code> and plot your results. From your results, how do population size and genetic drift relate?</p>
</section>
<section id="problem-3-modifying-the-number-of-generations">
<h2>Problem 3 – Modifying the number of generations<a class="headerlink" href="#problem-3-modifying-the-number-of-generations" title="Permalink to this headline">#</a></h2>
<p>Returning to <code class="docutils literal notranslate"><span class="pre">n_flies</span> <span class="pre">=</span> <span class="pre">16</span></code>, run the simulation for 100 generations and plot your results. From your results, what role does time seem to play in genetic drift?</p>
</section>
<section id="problem-4-going-beyond-bag-of-alleles">
<h2>Problem 4 – Going beyond “bag of alleles”<a class="headerlink" href="#problem-4-going-beyond-bag-of-alleles" title="Permalink to this headline">#</a></h2>
<p>So far we have assumed that all 32 alleles can be drawn with equal probability to make the next generation, but this ignores the fact that the alleles actually exist in discrete individuals and they can’t be drawn quite that randomly. For this, we will construct a population as a list of lists, where each item in the list represents a fly, and the list of <code class="docutils literal notranslate"><span class="pre">0</span></code>s and <code class="docutils literal notranslate"><span class="pre">1</span></code>s represents which alleles that fly has. (<code class="docutils literal notranslate"><span class="pre">1</span></code> corresponds to the red allele and <code class="docutils literal notranslate"><span class="pre">0</span></code> corresponds to the white allele). The initial Buri population will look like this:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># 16 orange flies, each with a red (1) and white (0) aelele </span>
<span class="n">pop</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mi">1</span><span class="p">,</span><span class="mi">0</span><span class="p">]]</span> <span class="o">*</span> <span class="n">n_flies</span><span class="p">)</span>
<span class="n">pop</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>array([[1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0],
       [1, 0]])
</pre></div>
</div>
</div>
</div>
<p>Model this population over time, taking the following steps:</p>
<ul class="simple">
<li><p>Pick two flies at random from the population</p></li>
<li><p>From each fly, pick an allele at random</p></li>
<li><p>Use these two alleles to generate a new fly and save it as part of a new population</p></li>
<li><p>Repeat this process <code class="docutils literal notranslate"><span class="pre">n_flies</span></code> times until you have produced a new population of 16 files</p></li>
</ul>
<p>With this new population in hand, you can record the new frequency of the red allele and repeat the whole process again for 20 generations. Plot the frequency of the red allele over time for this vial. For now we won’t worry about simulating this for 107 vials.</p>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
            </main>
            <footer class="footer-article noprint">
                
    <!-- Previous / next buttons -->
<div class='prev-next-area'>
    <a class='left-prev' id="prev-link" href="10_master_equation_gene_exp.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Week 10: Master equations applied to gene expression</p>
        </div>
    </a>
</div>
            </footer>
        </div>
    </div>
    <div class="footer-content row">
        <footer class="col footer"><p>
  
    By Dr. Suzannah Beeler<br/>
  
      &copy; Copyright 2022.<br/>
</p>
        </footer>
    </div>
    
</div>


      </div>
    </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="_static/scripts/pydata-sphinx-theme.js?digest=1999514e3f237ded88cf"></script>


  </body>
</html>